### PR TITLE
Zero dim Tensor support

### PR DESCRIPTION
**Pull request by Joqsan**
This PR adds the following features:

- Zero-dim support:
    
    ```python
    >>> from tinygrad.tensor import Tensor
    >>> b = Tensor(10)
    >>> b.shape
    ()
    >>> ones = Tensor.ones(2,2,2,2)
    >>> ones.sum().shape
    ()
    >>> (ones.sum() * b).numpy()
    array(160., dtype=float32)
    >>> (ones.sum() * b.reshape(1)).numpy()
    array([160.], dtype=float32)
    ```
    
- varargs support for `Tensor.zeros()` and `.ones()`: not only `Tensor.zeros(45, 65)` works, but also `Tensor.zeros([45, 65])`.
    - Why: Make them consistent with their torch counterparts: e.g., `Tensor.zeros([])` should return a 0dim tensor.
- A refactored `Tensor.__getitem__()`.
    - ~5.8x faster slicing:
        
        ```python
        d = torch.arange(64**3).reshape(64,64,64)
        %timeit old_getitem(d, s[1:32, 32:64, 1:64], print_results=False)
        %timeit new_getitem(d, s[1:32, 32:64, 1:64], print_results=False)
        23.6 µs ± 975 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
        4.06 µs ± 46.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
        
        %timeit old_getitem(d, s[1:32, 32:64, None, 1:64, None, None], print_results=False)
        %timeit new_getitem(d, s[1:32, 32:64, None, 1:64, None, None], print_results=False)
        28.5 µs ± 638 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
        4.86 µs ± 91.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
        ```
        
    - Remove the use of `.slice()`; it just relies on `.shrink()` and `.reshape()`.
    - Fix weird behaviors when slicing out of range:
        
        ```python
        >>> w = np.arange(8**3).reshape(8,8,8).astype(np.float32)
        >>> tg = Tensor(w)
        >>> pt = torch.from_numpy(w)
        >>> tg[-4:2].shape  # tensor of shape (-2, 8, 8)
        >>> pt[-4:2].shape  # tensor([], size=(0, 8, 8))
        
        ```
        
    - More torch-like exception messages if invalid slices passed.
    - (Also, it got easier to debug, which was the main reason to refactor it in the first place).

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -199,7 +198,7 @@ def backward(self):
 
   def reshape(self, shape, *args) -> Tensor:
     new_shape = argfix(shape, *args)
-    assert len(new_shape) > 0 and all(x != 0 for x in new_shape), f"zeros not allowed in shape {new_shape}"
+    assert len(new_shape) == 0 or all(x != 0 for x in new_shape), f"zeros not allowed in shape {new_shape}"
```
Comment (geohot) [inline tinygrad/tensor.py:201]: I don't think this first clause in the or is required.

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -215,37 +214,42 @@ def slice(self, arg:Sequence[Optional[Tuple[int, int]]]) -> Tensor:
     padding = tuple((max(0, -p[0]), max(0, p[1]-self.shape[i])) for i,p in enumerate(arg_))
     return self.pad(padding).shrink(tuple((p[0] + padding[i][0], p[1] + padding[i][0]) for i,p in enumerate(arg_)))
 
-  # Tensors mostly follow the normal python indexing / slicing behavior for sequences
```
Comment (geohot) [inline tinygrad/tensor.py:218]: Are these comments no longer true? I see you rewrote this function, are all the behaviors tested?

**Comment by geohot** [comment]
Very cool! A few comments, and a bunch of tests failing.

File: tinygrad/tensor.py (hunk 1 above)
Comment (Joqsan) [reply tinygrad/tensor.py:201]: Got it.

File: tinygrad/tensor.py (hunk 2 above)
Comment (Joqsan) [reply tinygrad/tensor.py:218]: They are true (I forgot to include a simpler version of the comments):
- Strides other than 1 raise an `AssertionError`.
- Empty slices are partially supported: forward pass works, but backward for tensors with shape containing zeros in needs to be revised, in order to make it work (idk who needs `.grad` of a zero-in-shape tensor).

The slicing behavior is preserved and tests too + I added some more tests from https://github.com/geohot/tinygrad/pull/506#discussion_r1093620465 and for zero-in-shape tensors.

**Comment by Joqsan** [comment]
> Very cool! A few comments, and a bunch of tests failing.

Yeah. Looking into it.

File: tinygrad/tensor.py (hunk 2 above)
Comment (geohot) [reply tinygrad/tensor.py:218]: Hmm, shape shouldn't ever actually contain 0s, but now you can have a 0 length shape.

Can you add a (failing) test for this unsupported behavior? It's fine if it's skipped.

Also, the strides should be easy to add support for. Pad+Reshape+Slice+Reshape

**Comment by geohot** [comment]
You interested in a job at the tiny corp by chance? Both this and your last PR were quite high quality.

File: tinygrad/tensor.py (hunk 2 above)
Comment (Joqsan) [reply tinygrad/tensor.py:218]: > Hmm, shape shouldn't ever actually contain 0s, but now you can have a 0 length shape. Is that what doesn't work backward?

Not sure if I understood your question, but backward doesn't work, regardless of 0dim tensors support or anything in this PR. 

I wanted to add support (maybe a separate PR) for a behavior with this kind of tensors similar to the one in PyTorch (e.g. if `pt` has shape, say, `[0, 3, 3]`, then `pt + 1` has also shape `[0, 3, 3]` and also allows to backpropagate towards leaf tensors).

On the other hand, who'd purposefully do bad slices on tensors, followed by some ops followed by backward? Maybe we should settle with just raising an exception in those cases for the moment.

> Can you add a (failing) test for this unsupported behavior? It's fine if it's skipped.

I'll add a commit with such a test.

> Also, the strides should be easy to add support for. Pad+Reshape+Slice+Reshape

I'll have a look at it.

**Comment by Joqsan** [comment]
I added a (skipped) test for the tensor with zero in shape. 

It looks like the problem lies not directly in backwards itself but in doing binary ops with them. It doesn't look that hard.

**Comment by Joqsan** [comment]
> You interested in a job at the tiny corp by chance? Both this and your last PR were quite high quality.

Yes, I think it'd be pretty cool, since 1) I'm looking for a job right now and 2) I like the project and  (+ giving einops support is haunting me xD). Please let me now if there is any chance of that.

**Comment by geohot** [comment]
Full time jobs are in person in San Diego, or we have remote task based contract jobs. Which are you interested in?

See https://github.com/geohot/tinygrad/discussions/788 for the hiring workflow.

**Comment by geohot** [comment]
Cool, finish this up and then think of which contract you want to tackle. There's the big 10k ones that should take 1-6 weeks, the 1k ones which are straightforward and can be done in 3-10 days, and the MLPerf model ones which are a day or two (these ones are bounties and not contracts).

List here: https://www.linkedin.com/feed/update/urn:li:activity:7066092360312500224/

**Comment by Joqsan** [comment]
> Cool, finish this up and then think of which contract you want to tackle. There's the big 10k ones that should take 1-6 weeks, the 1k ones which are straightforward and can be done in 3-10 days, and the MLPerf model ones which are a day or two (these ones are bounties and not contracts).
> 
> List here: https://www.linkedin.com/feed/update/urn:li:activity:7066092360312500224/

Pretty cool. I'll come back to this PR ASAP.

**Comment by geohot** [comment]
Putting a $200 locked bounty on finishing this up nicely and getting it merged.

**Comment by geohot** [comment]
A bunch of the failing onnx tests depend on this.

https://gist.github.com/geohot/546d7fded8df870f6e246b857091e331

**Comment by Joqsan** [comment]
> A bunch of the failing onnx tests depend on this.
> 
> https://gist.github.com/geohot/546d7fded8df870f6e246b857091e331

Working on this.

**Comment by Joqsan** [comment]
Done. PyTorch doesn't support indexing with negatives steps, so tests are done against numpy tensors.

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py             1               0
tinygrad/nn/image.py                       1               0
tinygrad/shape/shapetracker.py             2               2
tinygrad/tensor.py                        67              35
------------------------------------------------------------
total                                     71              37
------------------------------------------------------------
lines added in the tinygrad folder: 34
```

**Comment by geohot** [comment]
Congrats! E-mail me with how you'd like to be paid.
